caffe2/sgd/adagrad_op.h

```diff
@@ -81,6 +62,26 @@ class SparseAdagradOp final : public Operator<Context> {
         epsilon_(OperatorBase::GetSingleArgument<float>("epsilon", 1e-5f)) {}
 
   bool RunOnDevice() override {
+    // Enforce types
+    CAFFE_ENFORCE(OperatorBase::InputIsType<Tensor<Context>>(PARAM));
+    CAFFE_ENFORCE(OperatorBase::InputIsType<Tensor<Context>>(MOMENT_1));
+    CAFFE_ENFORCE(OperatorBase::InputIsType<Tensor<Context>>(INDICES));
+    CAFFE_ENFORCE(OperatorBase::InputIsType<Tensor<Context>>(GRAD));
+    CAFFE_ENFORCE(OperatorBase::InputIsType<Tensor<Context>>(LR));
+
+    // Enforce shapes
+    CAFFE_ENFORCE(Input(PARAM).size() == Input(MOMENT_1).size());
+    CAFFE_ENFORCE(Input(LR).size() == 1);
+    CAFFE_ENFORCE(Input(PARAM).size_from_dim(1) ==
+            Input(GRAD).size_from_dim(Input(INDICES).ndim()));
+
+    // These must be in-place for the sparse op. If out-of-place is required,
+    // we need to copy input to output before running.
+    CAFFE_ENFORCE_EQ(&Input(PARAM), Output(OUTPUT_PARAM),
```

Comment:
you can use `EnforceInplace` in the OpSchema definition in .cc file. Then you don't really need this. Also you can remove ResizeLike in the DoRunWithType in this case.